### PR TITLE
feat(protocol): Batch Reader

### PR DESCRIPTION
**Pull request by refcell**
### Description

Introduces the batch reader type.

Fixes broken batch decoding.

### CODE REVIEW

**Comment by refcell** [comment]
## 📚 $\text{Stack Overview}$

Pulls submitted in this stack:
* #234 👈
* `main`

_This comment was automatically generated by [`st`](https://github.com/clabby/st)._

File: crates/protocol/src/batch/transactions.rs [hunk 1]
```diff
@@ -260,11 +260,8 @@ impl SpanBatchTransactions {
                 .tx_sigs
                 .get(idx as usize)
                 .ok_or(SpanBatchError::Decoding(SpanDecodingError::InvalidTransactionData))?;
-            let is_protected = self
-                .protected_bits
-                .get_bit(idx as usize)
-                .ok_or(SpanBatchError::Decoding(SpanDecodingError::InvalidTransactionData))?
-                == 1;
+            let is_protected =
+                self.protected_bits.get_bit(idx as usize).unwrap_or(sig.v() as u8) == 1;
```
Comment (refcell) [inline crates/protocol/src/batch/transactions.rs:264]: @klkvr in your alloy update pr, span batch txs decoding broke when a tx didn't have protected bits - I've updated the `SpanBatchTransactions` handling here to use the signature v value if the protected bits are missing. This fixes batch reading. I've upstreamed the `BatchReader` from kona as well in this PR which validates batch reading - prior to this pr, decoding a valid batch with the `BatchReader` failed.

Comment (refcell) [reply crates/protocol/src/batch/transactions.rs:264]: For reference: https://github.com/anton-rs/kona/pull/786

Comment (klkvr) [reply crates/protocol/src/batch/transactions.rs:264]: looks like before `None` was treated as false already

https://github.com/alloy-rs/op-alloy/blob/edc82479bc6b0d2f48176f278638eaedf892b5f9/crates/protocol/src/batch/transactions.rs#L247-L250

File: crates/protocol/src/batch/reader.rs [hunk 2]
```diff
@@ -0,0 +1,128 @@
+//! Contains the [BatchReader] which is used to iteratively consume batches from raw data.
+
+use alloc::vec::Vec;
+use alloy_primitives::Bytes;
+use alloy_rlp::Decodable;
+use miniz_oxide::inflate::decompress_to_vec_zlib;
+use op_alloy_genesis::RollupConfig;
+
```
Comment (clabby) [inline crates/protocol/src/batch/reader.rs:8]:
```suggestion
```

File: crates/protocol/src/batch/transactions.rs (hunk 1 above)
Comment (clabby) [reply crates/protocol/src/batch/transactions.rs:264]: Yeah if the protected bit isn't found, we should be treating the signature as unprotected.

**Review by clabby — APPROVED** [review]
LGTM, pending protected bit fix.

File: crates/protocol/src/brotli.rs [hunk 3]
```diff
@@ -0,0 +1,105 @@
+//! Stage Utilities
+
+use alloc::{vec, vec::Vec};
+use alloc_no_stdlib::*;
```
Comment (mattsse) [inline crates/protocol/src/brotli.rs:4]: this seems weird, why do we need this?

**Review by mattsse — APPROVED** [review]
lgtm, but as discussed lets get rid of this dep

https://github.com/alloy-rs/op-alloy/blob/aac1566ed65015fbabafe5ba0bb49def6af3012d/crates/rpc-types-engine/Cargo.toml#L19-L19

**Comment by refcell** [comment]
Will come back to this once more batch type cleaning is done

**Comment by refcell** [comment]
Closing this for now - batch reading will strictly be used in the derivation pipeline so it shouldn't need to be used in other op stack components.

**Pull request closed without merging**